pkg/cmd/server/kubernetes/node_auth.go

```diff
@@ -0,0 +1,119 @@
+package kubernetes
+
+import (
+	"crypto/x509"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/auth/authenticator"
+
+	"k8s.io/kubernetes/pkg/auth/authorizer"
+	"k8s.io/kubernetes/pkg/auth/user"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+
+	oauthenticator "github.com/openshift/origin/pkg/auth/authenticator"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/bearertoken"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/unionrequest"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/x509request"
+	"github.com/openshift/origin/pkg/auth/authenticator/token/apitoken"
+	authncache "github.com/openshift/origin/pkg/auth/authenticator/token/cache"
+	"github.com/openshift/origin/pkg/auth/group"
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/authorization/authorizer/adapter"
+	authzcache "github.com/openshift/origin/pkg/authorization/authorizer/cache"
+	"github.com/openshift/origin/pkg/authorization/authorizer/subjectaccessreview"
+	oclient "github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+)
+
+func newAuthenticator(clientCAs *x509.CertPool, anonymousConfig client.Config) (authenticator.Request, error) {
+	authenticators := []oauthenticator.Request{}
+
+	// API token auth
+	tokenAuthenticator, err := apitoken.NewTokenAuthenticator(anonymousConfig)
+	if err != nil {
+		return nil, err
+	}
+	// Cache results
+	cachedTokenAuthenticator, err := authncache.NewAuthenticator(tokenAuthenticator, 5*time.Minute, 1000)
+	if err != nil {
+		return nil, err
+	}
+	authenticators = append(authenticators, bearertoken.New(cachedTokenAuthenticator, true))
+
+	// Client-cert auth
+	if clientCAs != nil {
```

Comment:
> if we set up an authenticator with nil verify Roots, I think you would

I'd like an issue opened to suggest that for the master and the kubelet.  Even though it would eliminate the ability to remove certificate auth, I might be ok with keying off of `clientCAs` being nil.  We don't play well without certificate auth.